### DISCUSSION TITLE
Are navigate, replace... functions identites stable?

### DISCUSSION THREAD

**Discussion by mfkrause** [react-navigation/react-navigation#10552] [Q&A] [answered]
Hey! I've recently stumbled upon a little hiccup when working with `@react-navigation/native` and hooks. Take this code as an example:

```js
const useSomeHook = () => {
  const navigation = useNavigation();
  const { navigate } = navigation;
  const [someState, setSomeState] = useState(someValue);

  useEffect(() => {
    if (someState === someValue) navigate('SomeScreen');
  }, [someState, navigate]);

  // ...
}
```

The effect should only trigger when `someState` has changed. Is this safe to do, as in, are the `navigate`, `replace`, etc. functions from the `useNavigation()` hook considered stable or can they change between renders? The only info on this I was able to find was [in an issue in react-navigation/hooks](https://github.com/react-navigation/hooks/issues/40) which has never been resolved, though (and react-navigation/hooks has since been deprecated as hooks are implemented in core already). I wasn't able to find any information on this in regards to the core hooks, though.
In order to satisfy the `exhaustive-deps` rule, I'm currently wrapping those functions in a `useRef()` to be sure. This adds quite a lot of boilerplate, though.

Would be grateful for any details on this!

**Comment by satya164** [comment, marked as answer]
They are stable, however, I'd suggest avoiding navigation in effects when possible. Navigating inside `onPress` (or other user interactions) provides the best user experience.

**Comment by mfkrause** [comment]
@satya164 Thanks for the quick answer! The example above was just a very minimal one, of course - in my use case, I'm listening to changes on the value provided by another hook which gets changed when the user opens the app from a deep link in another app. From one of those deep links, I want to navigate to a specific screen. When the user clicks the link, he fully expects to land on that screen. Might be a very specific use case though. But good to hear they are stable, I can therefore get rid of all the ref boilerplate. Thanks!
